File: Problem_04_palindrome/problem_4.py
import datetime

def is_palindrome(number):
    reversed_number = 0
    n = number

    while(n):
        reversed_number = reversed_number * 10 + (n % 10)
        #print("reversed_number:", reversed_number)
        n //= 10
        #print("n:", n)

    return number == reversed_number

time_start = datetime.datetime.now()

largest_palindrome = -1
for i in range(100, 1000):
    #print("i:", i)
    for j in range(100, 1000):
        product = i * j
        if is_palindrome(product) and product > largest_palindrome:
            largest_palindrome = product
            # print("Palindrome:", product)

time_end = datetime.datetime.now()

print("Result:", largest_palindrome)
print("Microseconds elapsed:", time_end - time_start)

File: Problem_04_palindrome/problem_4.rs
use std::time::SystemTime;

fn is_palindrome(input_number: i64) -> bool {
    let mut number = input_number;
    let mut reversed_number = 0;

    while number > 0 {
        let digit = number % 10;
        number = number / 10;

        reversed_number *= 10;
        reversed_number += digit;
    }

    input_number == reversed_number
}

fn main() {
    let mut largest_palindrome = -1;

    let time_start = SystemTime::now();

    for i in 100..1000 {
        for j in 100..1000 {
            if i < j {  // we have searched for that in i*j (this is j*i)
                break
            }

            let product = i * j;
            let is_product_palindrome = is_palindrome(product);
            if is_product_palindrome && product > largest_palindrome {
                // println!("{} * {} = {}, palindrome?: {}", i, j, product, is_product_palindrome);
                largest_palindrome = product;
            }
        }
    }

    match time_start.elapsed() {
        Ok(elapsed) => {
            println!("Calc duration: {}", elapsed.as_secs());
            println!("Subsec nanos: {}", elapsed.subsec_nanos());
        }
        Err(e) => {
            println!("Error: {:?}", e);
        }
    }

    println!("Largest 3-digit palindrome: {}", largest_palindrome);
}
